xonsh/ptk/history.py

```diff
@@ -77,9 +77,10 @@ def run(self):
                 continue
 
     def _buf(self):
-        if hasattr(builtins, '__xonsh_shell__'):
-            buf = builtins.__xonsh_shell__.shell.prompter.cli.application.buffer
-        else:
-            buf = None
+        # Thread-safe version of
+        # buf = builtins.__xonsh_shell__.shell.prompter.cli.application.buffer
+        buf = getattr(getattr(getattr(getattr(getattr(getattr(builtins, 
```

Comment:
Instead of chaining these together, could we loop over these things?

```
buf = builtins
for i in ['__xonsh_shell__', 'shell', 'prompter', 'cli', 'application', 'buffer']:
    buf = getattr(buf, i, None)
    if buf is None:
        break
return buff
```

It might not be as efficient in the average case (actually, I'm not sure), but it feels cleaner to me.

If you disagree, I'm happy to merge in as-is.